### PR TITLE
[WIP] open-mpi: fix Fortran support

### PR DESCRIPTION
**Pull request by UniqMartin**
**PLEASE DON'T MERGE!** Some parts shamelessly stolen from #46461.

### CODE REVIEW

File: Library/Formula/open-mpi.rb [hunk 1]
```diff
@@ -27,7 +29,7 @@ class OpenMpi < Formula
   conflicts_with "lcdf-typetools", :because => "both install same set of binaries."
 
   depends_on :java => :build
-  depends_on :fortran => :recommended
+  depends_on :fortran if build.with?("fortran")
```
Comment (xu-cheng) [inline Library/Formula/open-mpi.rb]: I would to fix the problem in the core since we repeatedly discourage others to do this.

Comment (mikemcquaid) [inline Library/Formula/open-mpi.rb]: Agreed :+1:

Comment (UniqMartin) [inline Library/Formula/open-mpi.rb]: I completely agree! As I said in the other PR, this is just a semi-ugly workaround and not the real solution. I myself would like to see all of those conditional `depends_on` eliminated (those within an `if`/`unless` block or with a trailing `if`/`unless` modifier). But fixing the issue here and eradicating conditional `depends_on` requires an in-depth understanding of Homebrew's dependency resolution logic – something I don't have (yet).

Comment (UniqMartin) [inline Library/Formula/open-mpi.rb]: This is really just a trick to move the evaluation to the formula, before the requirement has been transformed to a dependency with a different name, making the option handling in `brew install` fail to connect the `--with{,out}-fortran` (or its absence) with the `gcc` dependency.

This PR is really only here to check with the CI if the workaround has the desired effect.

**Comment by xu-cheng** [comment]
Looked around a little. This bug is originally from [`Dependency.merge_repeat`](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/dependency.rb#L121-L130). In which, the crucial  `option_name` is abandoned. In order to solve that, I think we may need to introduce `option_name_list` variable as one dependency would have multiple option names after merging. I will look further for solution when time is free.

**Comment by UniqMartin** [comment]
@xu-cheng If you can, please do. I agree that we need a proper solution to this issue eventually. For now, the dependency code is way to complicated for me to hack on (and to important to the project at large for me to break it).

**Comment by zbeekman** [comment]
Any chance #46553 is also caused by the same bug in Dependency.merge_repeat?

**Comment by UniqMartin** [comment]
@zbeekman Could be, but my gut feeling is there's something else going on there. I hope I'll be able to debug the mentioned issue soon (maybe later today or tomorrow) if no one else beats me to it.

**Pull request closed without merging**